src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -2174,43 +2198,18 @@ private AssemblySymbol GetForwardedToAssembly(string fullName, int arity, Diagno
                 while (current != null);
             }
 
-            // NOTE: This won't work if the type isn't using CLS-style generic naming (i.e. `arity), but this code is
-            // only intended to improve diagnostic messages, so false negatives in corner cases aren't a big deal.
-            var metadataName = MetadataTypeName.FromFullName(fullName, useCLSCompliantNameArityEncoding: true, forcedArity: arity);
-
-            var containingAssembly = this.Compilation.Assembly;
+            name = MetadataHelpers.ComposeAritySuffixedMetadataName(name, arity);
 
-            // This method is only called after lookup has failed, so the containing (source!) assembly can't
-            // have a forwarder to another assembly.
-            NamedTypeSymbol forwardedType = null;
-            foreach (var referencedAssembly in containingAssembly.Modules[0].GetReferencedAssemblySymbols())
+            var fullName = qualifierOpt != null && !ReferenceEquals(qualifierOpt, Compilation.GlobalNamespace) ? qualifierOpt.ToDisplayString(SymbolDisplayFormat.QualifiedNameOnlyFormat) + "." + name : name;
```

Comment:
>qualifierOpt.ToDisplayString(SymbolDisplayFormat.QualifiedNameOnlyFormat) [](start = 113, length = 73)

We don't normally use symbol display formatting for forming metadata names.  The code used to use `MetadataTypeName.FromFullName` for that.  Please restore that technique, along with the comment explaining its limitations.

Reply:
I'm not sure I understand your request here (sorry, I'm not at all familiar with the roslyn codebase, just trying to figure out a solution to a bug which is blocking us).

I'm trying to compose a namespace-qualified name to test different "using"s for existing forwarders. MetadataTypeName.FromFullName does not seem to add a namespace? Should I use MetadataTypeName FromNamespaceAndTypeName? 

And as for restoring the comment:
"NOTE: This won't work if the type isn't using CLS-style generic naming (i.e. `arity)"

But in my case I moved the composition of `arity into this function (var metadataName = MetadataHelpers.ComposeAritySuffixedMetadataName(name, arity);) in the line just above this one - so isn't it safe to assume that we always use `arity naming?